File: LeetCode/Python/Easy/938-Range-Sum-of-BST/solution.py
class Solution:
    def __init__(self):
        self.result=0

    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
        self.helper(root, low, high)

        return self.result

    def helper(self, root: Optional[TreeNode], low: int, high: int):
        if not root:
            return

        if low <= root.val and root.val <= high:
            self.result+=root.val

        self.helper(root.left, low, high)

        self.helper(root.right, low, high)

class Solution:
    def __init__(self):
        self.result=0

    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
        self.helper(root, low, high)

        return self.result

    def helper(self, root: Optional[TreeNode], low: int, high: int):
        if not root:
            return

        if low <= root.val and root.val <= high:
            self.result+=root.val

            self.helper(root.left, low, high)

            self.helper(root.right, low, high)
        elif root.val < low:
            self.helper(root.right, low, high)
        elif root.val > high:
            self.helper(root.left, low, high)

File: LeetCode/Rust/Easy/938-Range-Sum-of-BST/solution.rs
use std::rc::Rc;
use std::cell::RefCell;
type OptionNode=Option<Rc<RefCell<TreeNode>>>;

impl Solution
{
    pub fn range_sum_bst(root: Option<Rc<RefCell<TreeNode>>>, low: i32, high: i32) -> i32
    {
        let mut result: i32=0;

        Self::helper(&root, &mut result, low, high);

        result
    }

    fn helper(root: &OptionNode, result: &mut i32, low: i32, high: i32) -> ()
    {
        if root.is_none()
        {
            return ();
        }

        let node: &Rc<RefCell<TreeNode>>=root.as_ref().unwrap();

        let node_val: i32=node.borrow().val;

        if low <= node_val && node_val <= high
        {
            *result+=node_val;

            Self::helper(&node.borrow().left, result, low, high);

            Self::helper(&node.borrow().right, result, low, high);
        }
        else if node_val < low
        {
            Self::helper(&node.borrow().right, result, low, high);
        }
        else if node_val > high
        {
            Self::helper(&node.borrow().left, result, low, high);
        }
    }
}